### PR TITLE
Change get_standing_orders_of_slot to return always all standing_orders, including non-placed ones

### PR DESCRIPTION
**Pull request by josojo**
tackels #152 

testplan:
unit tests only

### CODE REVIEW

File: driver/src/models/standing_order.rs [hunk 1]
```diff
@@ -19,19 +19,17 @@ impl StandingOrder {
     pub fn get_orders(&self) -> &Vec<super::Order> {
         &self.orders
     }
+    pub fn get_orders_length(&self) -> usize {
```
Comment (bh2smith) [inline driver/src/models/standing_order.rs]: could we call this `num_order`? Then using it would be `standing_order.num_orders()`

File: driver/src/db_interface.rs [hunk 2]
```diff
@@ -124,12 +124,21 @@ impl DbInterface for MongoDB {
...
             .map(|d| d.map(models::StandingOrder::from).map_err(DriverError::from))
-            .collect::<Result<Vec<_>, _>>()
+            .collect::<Result<Vec<_>, _>>()?;
+
+        let empty_standing_order = models::StandingOrder::new(
+            0, 0, vec![]
+        );
+        let mut standing_orders = vec![empty_standing_order; models::NUM_RESERVED_ACCOUNTS as usize];
+        for standing_order in non_zero_standing_orders {
+            standing_orders[standing_order.account_id as usize] = standing_order.clone();
```
Comment (josojo) [inline driver/src/db_interface.rs]: I tried to be less clonely(), but did not find a way to do it. I have seen people switching to references or using std::mem::swap. But I think neither of these methods is appropriated here.

File: driver/src/models/standing_order.rs (hunk 1 above)
Comment (josojo) [reply driver/src/models/standing_order.rs]: yeah, that is better...

File: driver/src/db_interface.rs [hunk 3]
```diff
@@ -124,12 +124,21 @@ impl DbInterface for MongoDB {
...
-        self.client
+        let non_zero_standing_orders = self.client
             .db(models::DB_NAME)
             .collection("standing_orders")
             .aggregate(pipeline, None)?
             .map(|d| d.map(models::StandingOrder::from).map_err(DriverError::from))
-            .collect::<Result<Vec<_>, _>>()
+            .collect::<Result<Vec<_>, _>>()?;
+
+        let empty_standing_order = models::StandingOrder::new(
```
Comment (fleupold) [inline driver/src/db_interface.rs]: How about implementing `StandingOrder::empty()` ?

File: driver/src/db_interface.rs (hunk 2 above)
Comment (fleupold) [reply driver/src/db_interface.rs]: Does foreach work (https://docs.rs/foreach/0.3.0/foreach/)?

> ForEach trait and for_each! macro allow you to use iterator inside iteration loop, which is not posible when using for-in loop.

File: driver/src/models/standing_order.rs [hunk 4]
```diff
@@ -19,19 +19,17 @@ impl StandingOrder {
...
        let mut hasher = Sha256::new();
         hasher.input(init_hash);
         for i in 0..models::NUM_RESERVED_ACCOUNTS {
-            hasher.input(self
-                .iter()
-                .position(|o| o.account_id == i as u16) 
-                .map(|k| self[k].get_orders())
-                .map(|o| o.rolling_hash(0))
-                .unwrap_or(H256::zero()));
+            hasher.input(self[i as usize].get_orders().rolling_hash(0));
```
Comment (fleupold) [inline driver/src/models/standing_order.rs]: nit: 

```rust
for standing_order in self {
   hasher.input(standing_order.get_orders().rolling_hash(0));
}
```

File: driver/src/db_interface.rs (hunk 3 above)
Comment (fleupold) [reply driver/src/db_interface.rs]: Also, accountId shouldn't be always 0 but increase with each index, shouldn't it?

File: driver/src/order_driver.rs [hunk 5]
```diff
@@ -97,10 +98,9 @@ fn update_balances(state: &mut State, orders: &[Order], solution: &Solution) {
 }
 
 fn batch_index_from_standing_orders(standing_orders: &Vec<models::StandingOrder>) -> Vec<U128> {
-    let mut standing_order_indexes = vec![U128::zero(); models::NUM_RESERVED_ACCOUNTS as usize];
-
+    let mut standing_order_indexes = Vec::<U128>::new();
```
Comment (fleupold) [inline driver/src/order_driver.rs]: nit: this can become a 1-liner

```rust
standing_orders.iter().map(|o| U128::from(o.batch_index)).collect()
```

File: driver/src/db_interface.rs (hunk 2 above)
Comment (josojo) [reply driver/src/db_interface.rs]: I use for_each now. It makes the code easier to understand. However, I was not able to remove the clone.

File: driver/src/db_interface.rs [hunk 6]
```diff
@@ -124,12 +124,18 @@ impl DbInterface for MongoDB {
...
             .collection("standing_orders")
             .aggregate(pipeline, None)?
             .map(|d| d.map(models::StandingOrder::from).map_err(DriverError::from))
-            .collect::<Result<Vec<_>, _>>()
+            .collect::<Result<Vec<_>, _>>()?;
+
+        let mut standing_orders = (0..models::NUM_RESERVED_ACCOUNTS)
+            .map(|k| models::StandingOrder::empty(k as u16))
+            .collect::<Vec<models::StandingOrder>>();
+        non_zero_standing_orders.iter().for_each(|k| standing_orders[k.account_id as usize] = k.clone());
```
Comment (fleupold) [inline driver/src/db_interface.rs]: `into_iter` instead of `iter` will make it so that your `k` inside this closure is the value and not the reference. This way you don't need `k.clone()` anymore.

However, accessing the index (`k.account`) when setting an element of `Vec` is considered to happen after the value that is being set is consumed (in this case k, so k.account is accessing a value after it is consumed).
Therefore, you need to close k.account_id before using it (yet this is just cloning a single u32 instead of a full db record).

Here is the code:

```rust
non_zero_standing_orders.into_iter().for_each(|k| {
  let acc_id = k.account_id.clone() as usize;
  standing_orders[acc_id] = k;
});
```

**Comment by bh2smith** [comment]
Tests appear to be failing here because of maximum retires:

```
waited 181 seconds
Before retry #1: sleeping 0.3 seconds
Before retry #2: sleeping 0.6 seconds
Before retry #3: sleeping 1.2 seconds
Before retry #4: sleeping 2.4 seconds
Before retry #5: sleeping 4.8 seconds
Retries exhausted
The command "sh ./test/e2e-tests-standing-order.sh" exited with 1.
```

Maybe it is a good time for us to try and fix this.

**Comment by josojo** [comment]
I figured out that using :
`Result<[models::StandingOrder; models::NUM_RESERVED_ACCOUNTS as usize], DriverError>;`
is not a wise, as we can not differentiate nullified orders ( orders with just zeros as parameters) and non-placed orders at all. But for the rolling hash caclulation, it actually does make a difference, whether we have a non-placed order or a zero order.
Differentiating these two cases would require the introduction of a counter of orders placed. I think this is unjustified complexity. I will revert the latest changes on this PR.

**Comment by fleupold** [comment]
> But for the rolling hash caclulation, it actually does make a difference, whether we have a non-placed order or a zero order.

You are right. I wonder if this is something we need to change in the future as in the snark we will likely fill up unoccupied slots with 0s.

**Comment by josojo** [comment]
> You are right. I wonder if this is something we need to change in the future as in the snark we will likely fill up unoccupied slots with 0s.

Yeah, then many things might be needed to change. Also, our rolling hash of the normal orders would be required to be computed differently. It would always need to be a rolling hash of 500 orders, just x<500 would no longer be an option, as the snark would also always compute 500 rolling hashes.

**Comment by josojo** [comment]
@fleupold I think the PR is ready. Changing the rolling_hash construction should be done in another pr, if we decided to do it.
Could you review it, please?

**Comment by josojo** [comment]
I pushed the code for a potential solution for using arrays for standing orders.

Using arrays, which for types not implementing the Copy trait is a f... pain.
See the 
https://stackoverflow.com/questions/28656387/initialize-a-large-fixed-size-array-with-non-copy-types

I could not implement the macro solution, as rust does not implement a clean way to share a macro over several crates
I could not use the Default trait, as it would be required to be implemented in std::default::Default.
I can not write the 
```
impl Default for [StandingOrder; models::NUM_RESERVED_ACCOUNTS]
```
because only traits defined in the current crate can be implemented for arbitrary types.

Now, I came up with a solution, but it uses unsafe code.

File: driver/src/db_interface.rs [hunk 7]
```diff
@@ -116,20 +115,29 @@ impl DbInterface for MongoDB {
...
+    ) -> Result<[models::StandingOrder; models::NUM_RESERVED_ACCOUNTS], DriverError> {
         let pipeline = vec![
             doc!{"$match" => (doc!{"validFromAuctionId" => (doc!{ "$lte" => slot})})},
             doc!{"$sort" => (doc!{"validFromAuctionId" => -1, "_id" => -1})},
             doc!{"$group" => (doc!{"_id" => "$accountId", "orders" => (doc!{"$first" =>"$orders" }), "batchIndex" => (doc!{"$first" => "$batchIndex" })})}
         ];
 
-        info!("Querying standing_orders: {:?}", pipeline);
-        self.client
+               info!("Querying standing_orders: {:?}", pipeline);
```
Comment (fleupold) [inline driver/src/db_interface.rs]: nit: indentation

File: driver/src/db_interface.rs [hunk 8]
```diff
@@ -116,20 +115,29 @@ impl DbInterface for MongoDB {
     fn get_standing_orders_of_slot(
         &self,
         slot: u32,
-    ) -> Result<Vec<models::StandingOrder>, DriverError> {
+    ) -> Result<[models::StandingOrder; models::NUM_RESERVED_ACCOUNTS], DriverError> {
```
Comment (fleupold) [inline driver/src/db_interface.rs]: 👍

File: driver/src/models/standing_order.rs [hunk 9]
```diff
@@ -19,20 +19,26 @@ impl StandingOrder {
     pub fn get_orders(&self) -> &Vec<super::Order> {
         &self.orders
     }
+    pub fn num_orders(&self) -> usize {
+        self.orders.len()
+    }
+    pub fn empty(account_id: u16) -> StandingOrder {
+        models::StandingOrder::new(account_id, 0, vec![])
+    }
+    pub fn initialize_array(&self) -> [StandingOrder; models::NUM_RESERVED_ACCOUNTS] {
```
Comment (fleupold) [inline driver/src/models/standing_order.rs]: I liked the idea of having a macro that can be used for all kinds of types more than having `unsafe` specifically for StandingOrder, but I guess it's not a big issue. 

https://github.com/JoshMcguigan/arr_macro is a crate that has this feature claiming 100% safe rust code.

File: driver/src/db_interface.rs (hunk 7 above)
Comment (bh2smith) [reply driver/src/db_interface.rs]: I guess clippy doesn't catch these things....

File: driver/src/db_interface.rs [hunk 10]
```diff
@@ -116,24 +115,24 @@ impl DbInterface for MongoDB {
...
             .collection("standing_orders")
             .aggregate(pipeline, None)?
             .map(|d| d.map(models::StandingOrder::from).map_err(DriverError::from))
             .collect::<Result<Vec<_>, _>>()?;
 
-        let mut standing_orders = (0..models::NUM_RESERVED_ACCOUNTS)
-            .map(|k| models::StandingOrder::empty(u16::from(k)))
-            .collect::<Vec<models::StandingOrder>>();
+        let mut standing_orders = models::StandingOrder::empty(0).initialize_array();
+        (0..models::NUM_RESERVED_ACCOUNTS)
```
Comment (bh2smith) [inline driver/src/db_interface.rs]: I am not sure I understand this line here... Does it connect to the previous line? If so, then this appears to be an indentation issue, but if not what is it?

File: driver/src/models/standing_order.rs (hunk 9 above)
Comment (bh2smith) [reply driver/src/models/standing_order.rs]: Could you elaborate here on the meaning of `unsafe`?

File: driver/src/models/mod.rs [hunk 11]
```diff
@@ -1,6 +1,7 @@
 pub mod state;
 pub mod flux;
 pub mod order;
+#[macro_use]
```
Comment (bh2smith) [inline driver/src/models/mod.rs]: This appears out of nowhere? I also don't see any changes in this file that relate to the inclusion of this line. Was this intentional?

File: driver/src/models/standing_order.rs [hunk 12]
```diff
@@ -25,9 +25,16 @@ impl StandingOrder {
     pub fn empty(account_id: u16) -> StandingOrder {
         models::StandingOrder::new(account_id, 0, vec![])
     }
+    pub fn initialize_array(&self) -> [StandingOrder; models::NUM_RESERVED_ACCOUNTS] {
+        let mut array: [StandingOrder; models::NUM_RESERVED_ACCOUNTS] = unsafe { std::mem::uninitialized() };
```
Comment (bh2smith) [inline driver/src/models/standing_order.rs]: This seems... somehow unsafe.

File: driver/src/models/standing_order.rs (hunk 9 above)
Comment (bh2smith) [reply driver/src/models/standing_order.rs]: Never mind, I will read about it here --> [Unsafe Rust](https://doc.rust-lang.org/book/ch19-01-unsafe-rust.html)

Comment (josojo) [reply driver/src/models/standing_order.rs]: Well, generally, I would agree. But I am not so sure whether I am happy about pulling another package in there with an 0.* release. If we need it this initialization in two places, I would probably also use the package, but for just one place, I would stick with the unsafe code

I think having this small unsafe code is not much of an issue. Unsafe just means that the compiler can't run all the usual checks.

File: driver/src/db_interface.rs (hunk 10 above)
Comment (josojo) [reply driver/src/db_interface.rs]: It just creates an iterator for the numbers 0 to NUM_RESEVER_ACCOUNTS. Yeah it looks a little bit unsual

File: driver/src/models/mod.rs (hunk 11 above)
Comment (josojo) [reply driver/src/models/mod.rs]: How did you find this one? I had already pushed the cleaning commit a long time before your comment,

File: driver/src/models/standing_order.rs (hunk 9 above)
Comment (fleupold) [reply driver/src/models/standing_order.rs]: I don't think the version number is a good argument as they are completely arbitrary and have very little meaning.

> I think having this small unsafe code is not much of an issue.

It can easily be an argument next time that there is already a place with unsafe code so it's fine to add another (cf. [The Broken Window Theory](https://www.freecodecamp.org/news/pragmatic-programmer-broken-windows-6916998eecbe/))

I don't insist on changing it but wanted to point this out.

Comment (bh2smith) [reply driver/src/models/standing_order.rs]: Cute. I enjoyed reading about the _Broken Window Theory_

File: docker/driver/base/Dockerfile [hunk 13]
```diff
@@ -4,8 +4,8 @@ RUN apt-get update \
     && apt-get install -y --no-install-recommends curl ssh git libc-dev gcc openssl pkg-config libssl-dev \
     && rm -rf /var/lib/apt/lists/*
 
-#Install Rust
-RUN curl https://sh.rustup.rs -sSf | sh -s -- -y
+#Install Rust with version 1.36
```
Comment (fleupold) [inline docker/driver/base/Dockerfile]: How is this enforced? Doesn't it always install the newest rust version?

File: .travis.yml [hunk 14]
```diff
@@ -11,6 +11,8 @@ matrix:
         - rustup component add clippy
       script:
         - cd driver/
+        - ls /home/travis/.cargo/registry/src/
```
Comment (fleupold) [inline .travis.yml]: this doesn't seem to do anything.

File: .travis.yml [hunk 15]
```diff
@@ -11,6 +11,8 @@ matrix:
...
+        - echo "cognitive-complexity-threshold = 30" > '/home/travis/.cargo/registry/src/github.com-1ecc6299db9ec823/mongodb-0.3.12/clippy.toml'
```
Comment (fleupold) [inline .travis.yml]: Why is this needed?

File: docker/driver/base/Dockerfile (hunk 13 above)
Comment (josojo) [reply docker/driver/base/Dockerfile]: Yes it does. I tried to trigger a rebuild and only later realized that I have to push the docker image manully.
I will revert it

File: .travis.yml (hunk 14 above)
Comment (josojo) [reply .travis.yml]: was just for testing... sorry

**Comment by josojo** [comment]
I could not find a solution for the clippy issue:
https://github.com/rust-lang/rust-clippy/issues/3874

`--cap-lints allow` would not work, as we want warnings also to trigger a failure of test

--- locally I can just run:
`
        - echo "cognitive-complexity-threshold = 30" > ~/.cargo/registry/src/github.com-1ecc6299db9ec823/mongodb-0.3.12/clippy.toml  
`
and everything works. But this is of course not a good solution.
Also, because the file is created in the moment of running clippy, this does not work in Travis.

Does anyone have a suggestion?

File: .travis.yml [hunk 16]
```diff
@@ -11,7 +11,7 @@ matrix:
         - rustup component add clippy
       script:
         - cd driver/
-        - cargo clippy -- -D warnings
+        - cargo clippy -- cap-lints  # -- clippy has an issue: https://github.com/rust-lang/rust-clippy/issues/3874
```
Comment (bh2smith) [inline .travis.yml]: I saw that one of the suggestions in the thread here was to use `--cap-lints`. Could you mentioned a little more about what it does?

**Comment by bh2smith** [comment]
Hey @josojo - It appears there is a problem compiling a couple dependencies? Is this because of the rust version or clippy?

```
1.83s$ cargo clippy -- cap-lints
    Checking mongodb v0.3.12
    Checking tokio-uds v0.1.7
error: multiple input filenames provided (first two filenames are `/home/travis/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-uds-0.1.7/src/lib.rs` and `cap-lints`)
error: Could not compile `tokio-uds`.
warning: build failed, waiting for other jobs to finish...
error: multiple input filenames provided (first two filenames are `/home/travis/.cargo/registry/src/github.com-1ecc6299db9ec823/mongodb-0.3.12/src/lib.rs` and `cap-lints`)
error: Could not compile `mongodb`.
To learn more, run the command again with --verbose.
The command "cargo clippy -- cap-lints" exited with 101.
```

File: .travis.yml (hunk 16 above)
Comment (josojo) [reply .travis.yml]: I could really not find any good documentation of that.

If Clippy is linting all the code, it also needs to know the code of the dependencies, as it needs to infer some types. That is why Clippy is looking into the mongodb library, but in there the clippy.toml file is not valid with the new rust version. This is causing the issue. Cap-lints is supposed to cut off the linting so that especially we no longer have to compile the libraries. 

But obviously, it is not working and I can not find a solution right now.

Comment (fleupold) [reply .travis.yml]: They pushed a fix to the mongodb repo yesterday: https://github.com/mongodb-labs/mongo-rust-driver-prototype/commit/9203ec3216976cfbfddb07856b25d061965810af

Pinning the dependency to that commit should fix the issue.

Comment (josojo) [reply .travis.yml]: NICE!

Thanks a lot for unblocking

**Comment by bh2smith** [comment]
Just a question here: 
Does pinning to a specific commit of a repo mean that we should keep an eye open for the next release which includes these changes?

**Comment by josojo** [comment]
Yes, totally.
